### PR TITLE
remove fi_context usage

### PR DESCRIPTION
**Pull request by kseager**
not applicable to RMA fixes issue #285.

Signed-off-by: kseager <kayla.seager@intel.com>

### CODE REVIEW

File: src/transport_ofi.h [hunk 1]
```diff
@@ -240,13 +238,13 @@ void shmem_transport_ofi_drain_cq(void)
 
 		if(ret == 1) {
 			shmem_transport_ofi_frag_t *frag =
-				container_of(buf.op_context,
-				struct shmem_transport_ofi_frag_t,
-				context);
+                (shmem_transport_ofi_frag_t *) buf.op_context;
 
 			if(SHMEM_TRANSPORT_OFI_TYPE_BOUNCE == frag->mytype) {
+			    shmem_transport_ofi_bounce_buffer_t *bb =
```
Comment (sayantansur) [inline src/transport_ofi.h]: You could get rid of this by casting directly in the list_free macro?

File: src/transport_ofi.h [hunk 2]
```diff
@@ -460,8 +458,12 @@ static inline
 void
 shmem_transport_put_wait(long *completion) {
 
-    if((*completion)-- > 0)
+    shmem_internal_assertp((*completion) >= 0);
```
Comment (jdinan) [inline src/transport_ofi.h]: It should be fine to use a regular assertion here.  The ```p``` in ```assertp``` will cause it to persist even when error checking is disabled.

**Comment by kseager** [comment]
thanks for the feedback/review

**Comment by jdinan** [comment]
A note on this: ```fi_context``` was there to support the BG provider.  This should still work as long as we disable bounce buffering (so we don't generate CQ events).

**Comment by sayantansur** [comment]
Thanks for finding out why we added FI_CONTEXT in the first place!

**Comment by jdinan** [comment]
Does ```FI_CONTEXT``` require a ```struct fi_context``` in every operation, or just the operations that will generate CQ entries?  What I'm wondering is whether it's enough for us to detect ```FI_CONTEXT``` in ```fi_info:mode``` and disable bounce buffering (the only instance where we currently generate CQ entries), or whether we would have to generate CQ entries for every operation to manage ```struct fi_context``` objects at the middleware level?

**Comment by sayantansur** [comment]
It is required for operations that generate a completion entry. The application can remove the generation of completion entry by using the `FI_COMPLETION` flag (I think SHMEM is using that?). In that case the provider does not require `struct fi_context`.

The problem arises when there are two endpoints with differing `FI_CONTEXT` mode bits set that are attached to the same completion queue. When a completion is generated, the provider will get confused whether it can read the `context` parameter or not. Provider can't tell when a completion is generated due to FI_COMPLETION being a run time flag that can be controlled by the application. Therefore, this will be disallowed in ABI 1.1.

I'm not certain that detection of `FI_CONTEXT` is sufficient reason to disable bounce buffering, as they are really different things. Bounce buffering is used when memory copy bw > network draining bw. And FI_CONTEXT indicates whether or not the provider requires additional bits to track the transfer.

SHMEM can support `FI_CONTEXT` pretty easily since the only operations you require completions for (bounce buffers) have an easy place to store the struct.

I wonder if this should become its own compile time/configure time parameter?

**Comment by jdinan** [comment]
SOS _used_ to support FI_CONTEXT for the CQ events generated in the bounce buffering path.  This PR removed that support.

The stakes here are pretty low..  We only generate CQ events in the bounce buffering code (where we used to -- but as of this PR no longer -- support FI_CONTEXT).  If the provider requires FI_CONTEXT, we can easily achieve compatibility by disabling bounce buffering.  Or we can back out this PR.  I tend to favor just detecting the FI_CONTEXT requirement and disabling bounce buffering (most users want this anyway) when we bring up the OFI transport.

**Comment by sayantansur** [comment]
Yea, if that makes life easier, I'm all for it.

**Pull request merged**